### PR TITLE
include/cpp-btree: cleanups

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/include/cpp-btree/btree.h [hunk 1]
```diff
@@ -566,71 +522,54 @@ class btree_node {
...
+        int c = comp(key(mid), k);
+        if (c < 0) {
+          s = mid + 1;
+        } else if (c > 0) {
+          e = mid;
+        } else {
+          // Need to return the first value whose key is not less than k, which
+          // requires continuing the binary search. Note that we are guaranteed
+          // that the result is an exact match because if "key(mid-1) < k" the
+          // call to binary_search_compare_to() will return "mid".
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: updated and repushed.

**Comment by tchaikov** [comment]
please consider adapting https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/btree.h

**Comment by tchaikov** [comment]
the reason why i added the "performance" label is that, after this change, the containers using `pg_t` as the key of btree will be using the `pg_t::compare()` method for comparison. because it provides 3-way comparison, there is no need to do another round of search if `pg_t::compare()` returns 0. the same applies to other containers using arithmetic types as the key. these containers (`uint64_t_btree_t`) are used by bluestore.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-01-07_01:48:07-rados-wip-kefu-testing-2020-01-07-0032-distro-basic-smithi/

failures fixed / tracked by

- https://github.com/ceph/ceph/pull/32517
- https://github.com/ceph/ceph/pull/32518
- https://github.com/ceph/ceph/pull/32519
- https://tracker.ceph.com/issues/43431
- https://tracker.ceph.com/issues/43492
- https://tracker.ceph.com/issues/43494

File: src/include/cpp-btree/btree.h [hunk 2]
```diff
@@ -704,21 +653,21 @@ class btree_node {
...
+  void remove_value(const int i, allocator_type *alloc);
 
   // Removes the values at positions [i, i + to_erase), shifting all values
   // after that range to the left by to_erase. Does not change children at all.
   void remove_values_ignore_children(int i, int to_erase,
                                      allocator_type *alloc);
 
   // Rebalances a node with its right sibling.
-  void rebalance_right_to_left(int to_move, btree_node *right,
+  void rebalance_right_to_left(const int to_move, btree_node *right,
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: I'm not sure we need to reflect the constness at the declaration level.

File: src/include/cpp-btree/btree.h [hunk 3]
```diff
@@ -828,19 +762,19 @@ struct btree_iterator {
   using reference = Reference;
   using iterator_category = std::bidirectional_iterator_tag;
 
-  btree_iterator() : node(nullptr), position(-1) {}
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: OK, both fields are initialized immediately now.

File: src/include/cpp-btree/btree.h [hunk 4]
```diff
@@ -945,42 +876,43 @@ class btree {
     using field_type = typename node_type::field_type;
     node_type *parent;
     field_type position = 0;
-    field_type start = 0;
     field_type count = 0;
     // max_count must be != kInternalNodeMaxCount (so that this node is regarded
     // as a leaf node). max_count() is never called when the tree is empty.
     field_type max_count = node_type::kInternalNodeMaxCount + 1;
 
-#ifdef _MSC_VER
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: I guess we're not going need this. ;-)

File: src/include/cpp-btree/btree.h [hunk 5]
```diff
@@ -1449,18 +1409,8 @@ class btree {
...
-    return node_type::testonly_uses_linear_node_search();
-  }
-
  private:
-  // We use compressed tuple in order to save space because key_compare and
-  // allocator_type are usually empty.
-  absl::container_internal::CompressedTuple<key_compare, allocator_type,
-                                            node_type *>
-      root_;
+  empty_base_handle<key_compare, allocator_type, node_type*> root_;
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: ACK.

File: src/include/cpp-btree/btree.h [hunk 6]
```diff
@@ -945,42 +876,43 @@ class btree {
...
+    empty_base_handle(const Base1 &b1, const Base2 &b2, const Data &d)
+        : Base1(b1),
+          Base2(b2),
+          data(d) {}
+    Data data;
+  };
 
-    node_stats(size_type l, size_type i)
+  struct node_stats {
+    node_stats(ssize_t l, ssize_t i)
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: `ssize_t` was already mentioned before.

File: src/include/cpp-btree/btree.h [hunk 7]
```diff
@@ -1207,21 +1139,27 @@ class btree {
...
   template <typename K, typename LK>
   bool compare_keys(const K &x, const LK &y) const {
-    return compare_internal::compare_result_as_less_than(key_comp()(x, y));
+    return compare_result_as_less_than(key_comp()(x, y));
   }
 
-  value_compare value_comp() const { return value_compare(key_comp()); }
+  // Dump the btree to the specified ostream. Requires that operator<< is
+  // defined for Key and Value.
+  void dump(std::ostream &os) const {
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: That's a Ceph-specific extension, I bet.

File: src/include/cpp-btree/btree.h [hunk 8]
```diff
@@ -1300,29 +1238,40 @@ class btree {
...
   // Allocator routines.
-  allocator_type *mutable_allocator() noexcept {
-    return &root_.template get<1>();
+  allocator_type* mutable_allocator() noexcept {
+    return static_cast<allocator_type*>(&root_);
   }
-  const allocator_type &allocator() const noexcept {
-    return root_.template get<1>();
+  const allocator_type& allocator() const noexcept {
+    return *static_cast<const allocator_type*>(&root_);
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: ACK in the matter of the EBO part.

File: src/include/cpp-btree/btree.h [hunk 9]
```diff
@@ -1945,7 +1896,7 @@ auto btree<P>::insert_multi(const key_type &key, ValueType &&v) -> iterator {
 
 template <typename P>
 template <typename ValueType>
-auto btree<P>::insert_hint_multi(iterator position, ValueType &&v) -> iterator {
+auto btree<P>::insert_hint_multi(iterator position, ValueType &&v) ->iterator {
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: Typo? Why not to have the space here?

File: src/include/cpp-btree/btree_container.h [hunk 10]
```diff
@@ -40,10 +35,14 @@ class btree_container {
   // `key_arg<K>` evaluates to `K` when the functors are transparent and to
   // `key_type` otherwise. It permits template argument deduction on `K` for the
   // transparent case.
+  template <class Compare>
+  using is_transparent_t = typename Compare::is_transparent;
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree_container.h]: Do we really need this? I'm not 100% sure but it looks the only things with `is_transparent` defined were `StringBtreeDefaultLess` and `StringBtreeDefaultGreater`. We dropped both of them.

File: src/include/cpp-btree/btree.h [hunk 11]
```diff
@@ -42,73 +42,32 @@
...
-// A helper class that indicates if the Compare parameter is a key-compare-to
-// comparator.
+namespace btree::internal {
+
 template <typename Compare, typename T>
 using btree_is_key_compare_to =
-    std::is_convertible<absl::result_of_t<Compare(const T &, const T &)>,
-                        absl::weak_ordering>;
-
-struct StringBtreeDefaultLess {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: yes. i think it'd be more generic to detect `.compare()` than specializing for `string` and `string_view`.

File: src/include/cpp-btree/btree.h [hunk 12]
```diff
@@ -157,88 +133,87 @@ struct common_params {
 
   using allocator_type = Alloc;
   using key_type = Key;
-  using size_type = std::make_signed<size_t>::type;
+  using size_type = ssize_t;
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: sure. will revert to the original version.

File: src/include/cpp-btree/btree.h (hunk 2 above)
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: i just want to make it consistent.

Comment (rzarzynski) [reply src/include/cpp-btree/btree.h]: OK, I'm fine with that.

File: src/include/cpp-btree/btree.h (hunk 7 above)
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: yes. i intend to keep it in another PR. as this change is supposed to be backward compatible from the feature set perspective.

File: src/include/cpp-btree/btree.h [hunk 13]
```diff
@@ -157,88 +133,87 @@ struct common_params {
...
-
-    // Upper bound for the available space for values. This is largest for leaf
-    // nodes, which have overhead of at least a pointer + 4 bytes (for storing
-    // 3 field_types and an enum).
-    kNodeValueSpace =
-        TargetNodeSize - /*minimum overhead=*/(sizeof(void *) + 4),
+    kValueSize = ValueSize,
+    // Available space for values.  This is largest for leaf nodes,
+    // which has overhead no fewer than two pointers.
+    kNodeValueSpace = TargetNodeSize - 2 * sizeof(void*),
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: yes, i was using the old evaluation from the original version. the newer version of B+ tree implementation is more spatial efficient. i updated the comment to be in-sync with the adapted version. please re-review this part.

File: src/include/cpp-btree/btree.h (hunk 6 above)
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: fixed.

File: src/include/cpp-btree/btree_container.h (hunk 10 above)
Comment (tchaikov) [reply src/include/cpp-btree/btree_container.h]: `is_transparent` is a part of C++14 standard. and is respected by associate containers for heterogeneous comparison lookup. see https://en.cppreference.com/w/cpp/container/set/count . btree_container is trying to follow this convention for supporting this feature. see also http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3657.htm.

File: src/include/cpp-btree/btree.h (hunk 9 above)
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: fixed.

File: src/include/cpp-btree/btree.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/include/cpp-btree/btree.h]: Looks good. Grepping over the `.diff` doesn't show any introduction of `ssize_t` anymore.

File: src/include/cpp-btree/btree.h (hunk 7 above)
Comment (rzarzynski) [reply src/include/cpp-btree/btree.h]: My guess is that we will need tit sooner or later for debugging seastore. Let's keep.

File: src/include/cpp-btree/btree.h [hunk 14]
```diff
@@ -34,730 +34,827 @@
...
     kTargetNodeSize = TargetNodeSize,
 
-    // Available space for values.  This is largest for leaf nodes,
-    // which has overhead no fewer than two pointers.
-    kNodeValueSpace = TargetNodeSize - 2 * sizeof(void*),
+    // Upper bound for the available space for values. This is largest for leaf
+    // nodes, which have overhead of at least a pointer + 4 bytes (for storing
+    // 3 field_types and an enum).
+    kNodeValueSpace =
+        TargetNodeSize - /*minimum overhead=*/(sizeof(void *) + 4),
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: OK, the upgrade commit lowers the overhead while...

File: src/include/cpp-btree/btree.h [hunk 15]
```diff
@@ -163,82 +139,84 @@ struct common_params {
...
-
+    kValueSize = ValueSize,
     // Upper bound for the available space for values. This is largest for leaf
-    // nodes, which have overhead of at least a pointer + 4 bytes (for storing
-    // 3 field_types and an enum).
+    // nodes, which have overhead of at least a pointer + 3 bytes (for storing
+    // 3 field_types) + paddings. if alignof(key_type) is 1, the size of padding
+    // would be 0.
     kNodeValueSpace =
         TargetNodeSize - /*minimum overhead=*/(sizeof(void *) + 4),
```
Comment (rzarzynski) [inline src/include/cpp-btree/btree.h]: ... the customization commit touches only the comment. LGTM.

File: src/include/cpp-btree/btree.h [hunk 16]
```diff
@@ -1714,157 +1758,237 @@ void btree_iterator<N, R, P>::decrement_slow() {
...
+    insert_hint_unique(end(), params_type::key(*b), *b);
   }
 }
 
-template <typename P> template <typename ValuePointer>
-typename btree<P>::iterator
-btree<P>::insert_multi(const key_type &key, ValuePointer value) {
+template <typename P>
+template <typename ValueType>
+auto btree<P>::insert_multi(const key_type &key, ValueType&& v) ->iterator {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/include/cpp-btree/btree.h]: fixed and repushed.

**Comment by tchaikov** [comment]
jenkins test docs

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-01-16_04:08:27-rados-wip-kefu-testing-2020-01-15-2336-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-01-16_13:11:44-rados-wip-kefu-testing-2020-01-15-2336-distro-basic-smithi/

failures are not related.

**Pull request merged**